File: 0901/퍼즐 조각 채우기.py
from collections import deque, defaultdict
def solution(game_board, table):
    answer = 0
    
    N = len(game_board)

    blocks = defaultdict(list)

    visited = [[False for _ in range(N)] for _ in range(N)]
    # 블록 찾기
    for i in range(N):
        for j in range(N):
            if table[i][j] == 1 and not visited[i][j]:
                block = bfs(table, visited, i, j, 1)
                blocks[len(block)].append(standard(block))
                
    # 빈 공간 찾기
    for i in range(4):
        game_board = rotate(game_board)
        visited = [[False for _ in range(N)] for _ in range(N)]
        for i in range(N):
            for j in range(N):
                if game_board[i][j] == 0 and not visited[i][j]:
                    spce = bfs(game_board, visited, i, j, 0)
                    space = standard(spce)

                    len_space = len(space)
                    if blocks[len_space]:
                        for idx, block in enumerate(blocks[len_space]):
                            if space == block:
                                blocks[len_space].pop(idx)
                                answer += len_space
                                for r, c in spce:
                                    game_board[r][c] = 1
                                break

    return answer

def bfs(B, visited, r, c, target): 
    N = len(B)
    q = deque()
    # visited = [[False for _ in range(N)] for _ in range(N)]
    dir = [(0,1), (0,-1), (1,0), (-1,0)]
    visited[r][c] = True
    result = [[r,c]]
    
    q.append([r,c])
    while q:
        cr, cc = q.popleft()
        
        for dr, dc in dir:
            nr, nc = cr+dr, cc+dc
            if 0 <= nr < N and 0 <= nc < N:
                if B[nr][nc] == target and not visited[nr][nc]:
                    q.append([nr,nc])
                    visited[nr][nc] = True
                    result.append([nr,nc])
    
    return result

def standard(R): # 좌표계 변환
    R = sorted(R)
    std_x = R[0][0]
    std_y = R[0][1]
    tmp = []
    for x, y in R:
        tmp.append([x-std_x, y-std_y])
    
    return sorted(tmp)

def rotate(B): # 시계방향으로 90도 회전
    N = len(B)
    
    m = [[0 for _ in range(N)] for _ in range(N)]
    
    for r in range(N):
        for c in range(N):
            m[c][N-1-r] = B[r][c]
            
    return m

File: 0901/퍼즐 조각 채우기.js
function solution(game_board, table) {
  var answer = 0;
  const n = game_board.length;
  const board = game_board.map(v => v.slice())
  let blankList = [];
  for (let i = 0; i < n; i++){
      for (let j = 0; j < n; j++) {
          let blank = [];
          let queue = [];
          if (game_board[i][j] === 0) {
              queue.push([i, j]);
              while(queue.length !== 0) {
                  const [x, y] = queue.pop();
                  blank.push([x, y]);
                  game_board[x][y] = 1;
                  if (y < n-1 && game_board[x][y+1] === 0) queue.push([x, y+1]);
                  if (x < n-1 && game_board[x+1][y] === 0) queue.push([x+1, y]);
                  if (y > 1 && game_board[x][y-1] === 0) queue.push([x, y-1]);
                  if (x > 1 && game_board[x-1][y] === 0) queue.push([x-1, y]);
              }
              blankList.push(blank);
          }
      }
  }

  let blockList = [];
  for (let i = 0; i < n; i++){
      for (let j = 0; j < n; j++) {
          let block = [];
          let queue = [];
          if (table[i][j] === 1) {
              queue.push([i, j]);
              while(queue.length !== 0) {
                  const [x, y] = queue.pop();
                  block.push([x, y]);
                  table[x][y] = 0;
                  if (y < n-1 && table[x][y+1] === 1) queue.push([x, y+1]);
                  if (x < n-1 && table[x+1][y] === 1) queue.push([x+1, y]);
                  if (y > 1 && table[x][y-1] === 1) queue.push([x, y-1]);
                  if (x > 1 && table[x-1][y] === 1) queue.push([x-1, y]);
              }
              blockList.push(block);
          }
      }
  }

  const blanks = new Map();
  const blocks = new Map();
  for(let i = 1; i <= 6; i++) {
      blanks.set(i, []);
      blocks.set(i, []);
  }
  blankList.map(blank => {
      const tempList = blanks.get(blank.length);
      tempList.push(blank);
  });
  blockList.map(block => {
      const tempList = blocks.get(block.length);
      tempList.push(block);
  });
  for (let i = 1; i <= 6; i++) {
      let blockUsed = [];
      let blankUsed = [];
      let blankN = blanks.get(i);
      let blockN = blocks.get(i);
      if (blankN.length !== 0 && blockN.length !==0) {
          blankN.forEach((blank, idx) => {
              let isFill = false;
              for (let j = 0; j < blockN.length; j++) {
                  if (j in blockUsed) break;
                  const x = blockN[j][0][0];
                  const y = blockN[j][0][1];
                  
                  blockN[j].map(block => {
                      block[0] -= x;
                      block[1] -= y;
                  })
                  // 4번 돌려서 확인하기
                  for (let k = 0; k < 4; k++) {
                      blockN[j].map(block => {
                          const tmp = block[1]
                          block[1] = -block[0];
                          block[0] = tmp;
                      })
                      blank.map(start => {
                          let cnt = 0;
                          blockN[j].map(block => {
                              const x = start[0] + block[0];
                              const y = start[1] + block[1];
                              if (x >= 0 && y >= 0 && x < n && y < n && board[x][y] === 0) {
                                  cnt++;
                              }
                          })
                          if (cnt === i) {
                              isFill = true;
                              blockUsed.push(j);
                              blankUsed.push(idx);
                          }
                      })
                  }
              }
              if (isFill) {
                  answer += i
              }
          })
      }
  }
  return answer;
}